packages/core/src/awsService/ec2/prompter.ts

```diff
@@ -58,16 +58,19 @@ export class Ec2Prompter {
         return await client.getInstances()
     }
 
+    // TODO: implement a batched generator to avoid loading all instances into UI.
```

Comment:
For codecatalyst we use `AsyncCollection` , example: https://github.com/aws/aws-toolkit-vscode/blob/4e568970e8d8646cf8e590d39d8891d68cebde84/packages/core/src/shared/clients/codecatalystClient.ts#L427

and here is where that is used to build a prompter: https://github.com/aws/aws-toolkit-vscode/blob/4e568970e8d8646cf8e590d39d8891d68cebde84/packages/core/src/codecatalyst/wizards/selectResource.ts#L113-L118

Reply:
Looks like you already anticipated this, based on `packages/core/src/shared/clients/clientWrapper.ts` :)

Reply:
Yea, the issue is that we flatten the pages to process the instances (i.e. add name and status). So the two options would be either keep it unflattened (as Reservations https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/ec2/command/DescribeInstancesCommand/) or re-batch them. It seems preferable to keep them unflattened as possible, but this presents some changes elsewhere. Investigating this now as a followup.